src/environment.cpp

```diff
@@ -1256,6 +1256,50 @@ void ServerEnvironment::step(float dtime)
 		*/
 		removeRemovedObjects();
 	}
+
+	/*
+		Manage particle spawner expiration
+	*/
+	if (m_particle_management_interval.step(dtime, 1.0)) {
+		for(std::map<u32, float>::iterator i =
+		    m_particle_spawners.begin();
+		    i != m_particle_spawners.end();) {
+			//non expiring spawners
+			if (i->second <= -1.f) {
+				i++;
+				continue;
+			}
+
+			i->second -= 1.0;
+			if (i->second < 0.f)
+				m_particle_spawners.erase(i++);
+			else
+				i++;
```

Comment:
D'oh... see the discussion about this in the channel about this...

Both your version and Zeno's version have 2 rather serious bugs relating to iterators not getting incremented and problems if floating point arithmetic *doesn't* result in any floating point error.